### DISCUSSION TITLE
Collosal Cave...

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#123] [General]
I have some sources in c to a later version of "Collosal Cave Adventure", and am tempted to port it, but has it been already done? As ubiquitous as it is across architectures, I have to ask.

I think this version might be better than the DECUS version (Fortran & Macro-11), which although close to the original, isn't as expansive.

**Comment by kej715** [comment]
We do not yet have a working C compiler for NOS. A tape image containing
the UT C compiler exists, but no one has managed to use it successfully yet
to build an instance of the compiler. Thus, you would have a couple of
choices:

   1. Be the first to get the UT C compiler built successfully for NOS 2,
   then use it to port your original source code of the game.
   2. Manually translate the source code to a different language, e.g.,
   Pascal or FORTRAN, and port that.

cheers,
Kevin

**Reply by KC5VDJ** [reply]
I was going with option 2, thinking Pascal would be most suitable.  But if someone could find the UT c tape, it would be a decent addition, but our situation has been impeding spare time for a few months, but i would like to at least play with it.

Still, I'm thinking Pascal for the game though, and probably a bit quicker.

**Reply by wschaub** [reply]
CYBIL might be interesting then, it's pascal but enhanced with a lot of things you might expect from a C compiler. 

https://dl.acm.org/doi/pdf/10.1145/988327.988331

**Reply by KC5VDJ** [reply]
Well, we'll see...  Been looking at sympl in the system sources, sympl is similarly derived from Pascal with obvious lineage to it.

**Comment by kej715** [comment]
SYMPL is definitely not derived from Pascal. It predates Pascal and is a
language in the JOVIAL family.

**Reply by KC5VDJ** [reply]
My bad.  Never been exposed to jovial, so no reference point.

**Comment by pschow** [comment]
SYMPL predates Pascal and was derived from JOVIAL

**Reply by wschaub** [reply]
More than that it was part of the JOvial Compiler Implementation Tool it's the implementation language the jovial compiler was built on top of.   https://archive.org/details/DTIC_ADA061714

**Reply by KC5VDJ** [reply]
I've never actually touched jovial, so I was didn't have the reference point there...

**Comment by nickglazzard1959** [comment]
What an interesting document! I had thought SYMPL was a language specific
to CDC, but it also ran on Honeywell machines under GCOS. Well I never ...
Only time I've used it was to write a Sieve of Eratosthenes prime number
finder. Interesting language heavily used by NOS though (e.g. FSE and many
other things). CYBIL might be a better bet for new "systems" programs on
NOS 2, perhaps.

**Comment by philclaridge** [comment]
Returning to the start of the thread I wondered if the later versions of Adventure 5 or 6 (which are in Fortran) would be OK, and not need too much porting. See https://rickadams.org/adventure/e_downloads.html

Given that the data format does not change much between different versions of Adventure, if you have later game data than Adventure 5 or 6 you may want to simply back port any new game features into the existing Fortran.

Aside, not really a serious suggestion: How about a LLVM back end to target the CDC instruction set. Assuming enough memory (big if?) then you can have any language you want. Rust anyone? https://llvm.org/docs/WritingAnLLVMBackend.html

Phil

**Reply by wschaub** [reply]
does LLVM have the flexibility to be able to write a backend for a target with 1s compliment, no stack and no byte addressing? 
I'm genuinely curious if it's possible or not.

It would be interesting to play with writing a compiler written using modern tools that spits out COMPASS code for being assembled on NOS. I would like to learn a lot more about compilers in general even if I never made one that would count as a modern compiler with all the bells and whistles.

**Reply by KC5VDJ** [reply]
Go fishing for the old dragon book...saw it floating...and I must say that urs ammon's (et. al.) pascal-6000 is mighty pretty coding for the day, probably why everyone derived from it.  Not quite modern, but a great one to learn translators.  P-code would be closer to modern, representing intermediate codes.

**Reply by KC5VDJ** [reply]
They have to be more current than the DECUS pdp-11 Fortran / Macro-11...

**Comment by pschow** [comment]
>
> On Fri, Feb 9, 2024 at 2:20 PM William Schaub ***@***.***>
> wrote:
> >
> > does LLVM have the flexibility to be able to write a backend for a
> target with 1s compliment, no stack and no byte addressing?
> > I'm genuinely curious if it's possible or not.


I think the addressing would be the obstacle.  LLVM hard-codes the
assumption that bytes are individually addressable. The stack could
probably be simulated.

**Reply by KC5VDJ** [reply]
Wirth wrote on the caveats of negative zero being a rather sticky issue as well.

**Comment by kej715** [comment]
Also, can LLVM handle the case that operand registers are loaded or stored
as a side effect of loading other registers, and that the rules are not
symmetrical, i.e., X1 is loaded as a side effect of loading a value into
A1, but X6 is not loaded as a side effect of loading A6? Or, in order to
store a value from X2, you have to move it first to X6 or X7 and then set
A6 or A7, respectively. It's conceivable that rules for this behavior can
be defined, but it would probably be a bit complicated. It would probably
limit the useful register file to X1 - X5 as well.  Regarding stack, frame,
and heap pointers, those could easily be mapped to a couple of B registers.

FYI ... I've been working on a similar project ... using ACK (the Amsterdam
Compiler Kit) to create a C cross-compiler for the Cray X-MP. The problems
are similar, although the X-MP is a little less constrained. Like the Cyber
170, the Cray X-MP is word-addressable for data (it's parcel-addressable
for branching and subroutine calls). The ACK wants all target machines to
be byte-addressable. The mismatch is resolved by allowing the backend to
"pretend" that data is byte-addressable and then generating instructions to
shift and mask as needed to map "fake" byte addresses onto actual word
addresses and the bytes within them. The X-MP has a set of A registers that
are like Cyber 170 B registers, and it has S registers that are like X
registers. Fortunately, S registers are not loaded or stored as a side
effect of loading other registers. However, the X-MP registers do have
peculiarities of which the compiler backend needs to be aware, e.g.,
specifying S0 or A0 as an instruction operand can cause non-intuitive
things to happen. And, of course, the X-MP also has 64 scalar B registers,
8 vector V vector registers, and 64  vector T registers ...

cheers,
Kevin
